src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/object/ObjectKeysFunctionIterator.java

```diff
@@ -132,8 +132,16 @@ protected boolean hasNextLocal() {
 
     @Override
     protected void resetLocal(DynamicContext context) {
+        _alreadyFoundKeys = new ArrayList<>();
+        _nextResults = new LinkedList<>();
         _iterator.reset(_currentDynamicContext);
-        setResultsFromNextObjectItem();
+
+        if (_iterator.isDataFrame()) {
+            setResultsFromDF();
+        } else {
+            _iterator.open(_currentDynamicContext);
```

Comment:
There is still a redundancy with _iterator.reset() and _iterator.open(). A reset is logically equivalent to close+open, so that this line is in fact re-opening an already open iterator. Also, it is not necessary to reset _iterator if its DF API was used and will continue to be used. Maybe move reset to here and remove open?